### PR TITLE
Implemented Device parser

### LINKED ISSUE #592: Implement devices string parser
Relates to https://github.com/KostyaSha/yet-another-docker-plugin/issues/32

TODO add String parser for `Device.class` "/dev/sda:/dev/xvdc:r"

Parser
https://github.com/docker/docker/blob/6b4a46f28266031ce1a1315f17fb69113a06efe1/runconfig/opts/parse.go#L731-L764

Tests
https://github.com/docker/docker/blob/6b4a46f28266031ce1a1315f17fb69113a06efe1/runconfig/opts/parse_test.go#L468

### PR DESCRIPTION
**Pull request by KostyaSha**
fixes #592

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/593)

### CODE REVIEW

**Comment by KostyaSha** [comment]
Todo implement validator, but i have no time and daemon error would return error.

**Comment by KostyaSha** [comment]
@marcuslinke please review

File: src/main/java/com/github/dockerjava/api/model/Device.java [hunk 1]
```diff
@@ -45,6 +51,69 @@ public String getPathOnHost() {
...
+    }
+
+    /**
+     * ValidDeviceMode checks if the mode for device is valid or not.
+     * Valid mode is a composition of r (read), w (write), and m (mknod).
+     *
+     * @link https://github.com/docker/docker/blob/6b4a46f28266031ce1a1315f17fb69113a06efe1/runconfig/opts/parse.go#L796
+     */
+    private static boolean validDeviceMode(String deviceMode) {
+        Map<String, Boolean> validModes = new HashMap<>(3);
```
Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: The `validModes' map should be declared as static member and not declared inline.

File: src/test/java/com/github/dockerjava/api/model/DeviceTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ * @author Kanstantsin Shautsou
+ */
+public class DeviceTest {
+    @Test
+    public void testParse() throws Exception {
```
Comment (marcuslinke) [inline src/test/java/com/github/dockerjava/api/model/DeviceTest.java]: The test is missing the invalid mode case.

File: src/main/java/com/github/dockerjava/api/model/Device.java (hunk 1 above)
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: what do you mean under `not declared inline.`?

Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: I mean it should be declared at the class level not in the method itself.

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: It must be mutable, it changes true to false to exclude double letters.

File: src/main/java/com/github/dockerjava/api/model/Device.java [hunk 3]
```diff
@@ -45,6 +51,69 @@ public String getPathOnHost() {
...
+        if (isEmpty(deviceMode)) {
+            return false;
+        }
+
+        for (char ch : deviceMode.toCharArray()) {
+            final String mode = String.valueOf(ch);
+            if (isNotTrue(validModes.get(mode))) {
+                return false; // wrong mode
+            }
+            validModes.put(mode, false);
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: @marcuslinke ^

Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: What is it for? Some kind of optimization?

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: error duplicated modes i.e. `rrr`

File: src/main/java/com/github/dockerjava/api/model/Device.java [hunk 4]
```diff
@@ -45,6 +51,69 @@ public String getPathOnHost() {
...
+        validModes.put("m", true);
+
+        if (isEmpty(deviceMode)) {
+            return false;
+        }
+
+        for (char ch : deviceMode.toCharArray()) {
+            final String mode = String.valueOf(ch);
+            if (isNotTrue(validModes.get(mode))) {
+                return false; // wrong mode
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: if char repeats -> wrong mode

File: src/main/java/com/github/dockerjava/api/model/Device.java (hunk 3 above)
Comment (marcuslinke) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: OK. Now I got it. So it must be declared within the method then. Sorry for bothering.

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/Device.java]: yeah it tricky logic :)

**Comment by KostyaSha** [comment]
@marcuslinke updated, feel free to squash merge (dropdown in merge button) if it LGTY.

**Pull request merged**